### PR TITLE
Set replica count

### PR DESCRIPTION
**Pull request by empiredan**
https://github.com/apache/incubator-pegasus/issues/865

### CODE REVIEW

File: include/dsn/dist/replication/replication_ddl_client.h [hunk 1]
```diff
@@ -228,6 +228,12 @@ class replication_ddl_client
     error_with<query_app_manual_compact_response>
     query_app_manual_compact(const std::string &app_name);
 
+    error_with<configuration_get_max_replica_count_response>
+    get_max_replica_count(const std::string &app_name);
```
Comment (levy5307) [inline include/dsn/dist/replication/replication_ddl_client.h:232]: get_max_replica_count(const std::string &app_name) const;

**Comment by levy5307** [comment]
Could you add your test progress in description?

**Comment by empiredan** [comment]
> Could you add your test progress in description?

The progress of manual tests ?

File: src/meta/server_state.cpp [hunk 2]
```diff
@@ -3112,5 +3112,511 @@ void server_state::on_query_manual_compact_status(query_manual_compact_rpc rpc)
     response.__set_progress(total_progress);
 }
 
+void server_state::get_max_replica_count(configuration_get_max_replica_count_rpc rpc)
```
Comment (levy5307) [inline src/meta/server_state.cpp:3115]: void server_state::get_max_replica_count(configuration_get_max_replica_count_rpc rpc) const

File: src/meta/server_state.cpp [hunk 3]
```diff
@@ -3112,5 +3112,511 @@ void server_state::on_query_manual_compact_status(query_manual_compact_rpc rpc)
...
+                     server_state::sStateHash);
+}
+
+void server_state::update_app_max_replica_count(configuration_set_max_replica_count_rpc rpc)
+{
+    const auto &app_name = rpc.request().app_name;
+    const auto new_max_replica_count = rpc.request().max_replica_count;
+    auto &response = rpc.response();
+
+    zauto_read_lock l(_lock);
```
Comment (levy5307) [inline src/meta/server_state.cpp:3567]:
```
    zauto_read_lock l(_lock);

    GET_APP_AND_CHECK_IF_EXISTS(app_name);

    auto app_path = get_app_path(*app);

    auto ainfo = *(reinterpret_cast<app_info *>(app.get()));
    ainfo.max_replica_count = new_max_replica_count;
```
==>
```
{
    zauto_read_lock l(_lock);

    GET_APP_AND_CHECK_IF_EXISTS(app_name);

    auto app_path = get_app_path(*app);

    auto ainfo = *(reinterpret_cast<app_info *>(app.get()));
    ainfo.max_replica_count = new_max_replica_count;
}
```

File: src/meta/server_state.cpp [hunk 4]
```diff
@@ -3112,5 +3112,511 @@ void server_state::on_query_manual_compact_status(query_manual_compact_rpc rpc)
...
+             old_config_str,
+             new_config_str);
+}
+
+void server_state::update_next_partition_max_replica_count(
+    const app_state &app, int32_t partition_index, configuration_set_max_replica_count_rpc rpc)
+{
+    const auto &app_name = app.app_name;
+
+    const int32_t next_partition_index = partition_index + 1;
```
Comment (levy5307) [inline src/meta/server_state.cpp:3530]: const int32_t next_partition_index ==> const auto next_partition_index

File: src/meta/server_state.cpp [hunk 5]
```diff
@@ -3112,5 +3112,511 @@ void server_state::on_query_manual_compact_status(query_manual_compact_rpc rpc)
...
+
+    update_next_partition_max_replica_count(*app, partition_index, rpc);
+}
+
+void server_state::update_partition_max_replica_count_locally(
+    const partition_configuration &new_partition_config, app_state &app)
+{
+    const auto &gpid = new_partition_config.pid;
+    const auto app_id = gpid.get_app_id();
+    const auto partition_index = gpid.get_partition_index();
```
Comment (levy5307) [inline src/meta/server_state.cpp:3487]: Define these variables before you use it

**Comment by levy5307** [comment]
> > Could you add your test progress in description?
> 
> The progress of manual tests ?

Yes

**Comment by empiredan** [comment]
> include/dsn/dist/replication/replication_ddl_client.h

OK

**Comment by empiredan** [comment]
> > > Could you add your test progress in description?
> > 
> > 
> > The progress of manual tests ?
> 
> Yes

OK

File: src/meta/server_state.cpp [hunk 6]
```diff
@@ -3112,5 +3112,517 @@ void server_state::on_query_manual_compact_status(query_manual_compact_rpc rpc)
...
+    const auto new_max_replica_count = rpc.request().max_replica_count;
+
+    auto &response = rpc.response();
+    int32_t app_id = 0;
+
+    {
+        zauto_read_lock l(_lock);
+
+        auto app = get_app(app_name);
+        if (app == nullptr) {
```
Comment (levy5307) [inline src/meta/server_state.cpp:3151]: what about reuse `GET_APP_AND_CHECK_IF_EXISTS` here?

File: src/meta/server_state.cpp [hunk 7]
```diff
@@ -3112,5 +3112,517 @@ void server_state::on_query_manual_compact_status(query_manual_compact_rpc rpc)
...
+                new_max_replica_count);
+
+        tasking::enqueue(LPC_META_STATE_HIGH,
+                         tracker(),
+                         [this, rpc]() { update_app_max_replica_count(rpc); },
+                         server_state::sStateHash);
+        return;
+    }
+
+    auto &old_partition_config = app->partitions[partition_index];
```
Comment (levy5307) [inline src/meta/server_state.cpp:3270]: const auto &old_partition_config

File: src/meta/server_state.cpp (hunk 6 above)
Comment (empiredan) [reply src/meta/server_state.cpp:3151]: Good idea !

File: src/meta/server_state.cpp [hunk 8]
```diff
@@ -3112,5 +3112,517 @@ void server_state::on_query_manual_compact_status(query_manual_compact_rpc rpc)
...
+
+void server_state::update_next_partition_max_replica_count(
+    const app_state &app, int32_t partition_index, configuration_set_max_replica_count_rpc rpc)
+{
+    const auto &app_name = app.app_name;
+
+    const auto next_partition_index = partition_index + 1;
+    const auto partition_count = static_cast<int32_t>(app.partitions.size());
+
+    const auto new_max_replica_count = rpc.request().max_replica_count;
```
Comment (levy5307) [inline src/meta/server_state.cpp:3535]: move to L3538

File: src/common/meta_admin.thrift [hunk 9]
```diff
@@ -394,3 +394,27 @@ struct ddd_diagnose_response
...
+struct configuration_get_max_replica_count_request
+{
+    1:string                    app_name;
+}
+
+struct configuration_get_max_replica_count_response
+{
+    1:dsn.error_code            err;
+    2:i32                       max_replica_count;
+}
```
Comment (hycdong) [inline src/common/meta_admin.thrift:407]: I recommend adding `string hint_message` in this response.

**Comment by hycdong** [comment]
I have reviewed this pull request roughly, and I think this pull request is too big, you can seperate into at lease 2 pull request. For one, query table factor, for another one, update table factor.

Besides, As https://github.com/apache/incubator-pegasus/issues/865 shows, you have some code with updating table factor NOT in this pull request, and it seems your final design are not included in this issue, and the comments are not enough in your code. I didn't know some configurations, such as `max_reserved_number_of_dropped_replicas` and `INCREASE_MAX_REPLICA_COUNT_ONLY`.

In conclusion, I suggest that:
1. Seperate this pull request into at lease 2 pull requests
2. Fillful some comments in your code

File: src/meta/server_state.cpp [hunk 10]
```diff
@@ -3112,5 +3112,517 @@ void server_state::on_query_manual_compact_status(query_manual_compact_rpc rpc)
...
+    context.msg = nullptr;
+
+    auto new_partition_config = old_partition_config;
+    new_partition_config.max_replica_count = new_max_replica_count;
+    ++(new_partition_config.ballot);
+    context.pending_sync_task =
+        update_partition_max_replica_count_on_remote(new_partition_config, rpc);
+}
+
+task_ptr server_state::update_partition_max_replica_count_on_remote(
```
Comment (Shuo-Jia) [inline src/meta/server_state.cpp:3337]: You better comment on every function to point it run which thread pool, like: https://github.com/XiaoMi/rdsn/blob/v2.3.0/src/meta/meta_bulk_load_service.cpp#L48

**Comment by Shuo-Jia** [comment]
I suggest the willing next pr can be `little`,  large diff will make review difficult.  Since the pr has resolved most comment, I can ignore current pr

**Comment by acelyc111** [comment]
has been merged by other PRs.
